### ISSUE TITLE
Provide default locale for Translatable

### ISSUE THREAD

**Issue by charlie-wasp** [KnpLabs/DoctrineBehaviors#182]
Hello! I am searching for the way to set provide default locale for `translate` method. Unfortunately, I found out this in `TranslatableMethods`:

```
    public function getDefaultLocale()
    {
        return 'en';
    }
```

So, I guess, there is no way to override it? :smile:

**Comment by kuczek** [comment]
@charlieWasp If you are searching for default locale you probably should $translator->getFallbackLocales()[0] (with isset probably), and it also can be done like that here (i think it should) @docteurklein?

**Comment by kuczek** [comment]
Probably we should think how to use it with different language providers, in symfony it will be defaultFallback, and all fallbacks should be supported...

**Comment by markitosgv** [comment]
Im using in a Symfony app, and i think the best way is modify this method to read a configuration key to set this locale.

But its a trait, and i don´t know how we can configure this value, maybe a setter method?

**Comment by kuczek** [comment]
@markitosgv The bigest problem is, that this is not bundle, and not only symfony, so we need to prepare universal mechanism.

**Comment by charlieWasp** [comment]
@kuczek can we move default locale to parameter in `orm-services.yml`? So clients will be able to override it. Or maybe we can use callable class for it

**Comment by markitosgv** [comment]
@kuczek i know, maybe @charlieWasp proposal to move parameter to .yml is a good solution

**Comment by kuczek** [comment]
@markitosgv @charlieWasp But what about non Symfony usage? I agree that we should somehow make it settable, but maybe we should wrote/edit some listener that will inject this from somewhere. For symfony it will be the best solution, for the rest it will be possible to do that, and default value can stay as [en].

Maybe we should also think about creating proper fallback system, so if you do not have one translation, it's trying to get first fallback, second fallback, etc.

**Comment by markitosgv** [comment]
Maybe we could inser a third fallbackLocale argument in proxyCurrentLocaleTranslation?

```php
<?php

/*
 * This file is part of the KnpDoctrineBehaviors package.
 *
 * (c) KnpLabs <http://knplabs.com/>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Knp\DoctrineBehaviors\Model\Translatable;

use Doctrine\Common\Collections\ArrayCollection;

/**
 * Translatable trait.
 *
 * Should be used inside entity, that needs to be translated.
 */
trait TranslatableMethods
{
    /**
     * Returns collection of translations.
     *
     * @return ArrayCollection
     */
    public function getTranslations()
    {
        return $this->translations = $this->translations ?: new ArrayCollection();
    }

    /**
     * Returns collection of new translations.
     *
     * @return ArrayCollection
     */
    public function getNewTranslations()
    {
        return $this->newTranslations = $this->newTranslations ?: new ArrayCollection();
    }

    /**
     * Adds new translation.
     *
     * @param Translation $translation The translation
     *
     * @return $this
     */
    public function addTranslation($translation)
    {
        $this->getTranslations()->set($translation->getLocale(), $translation);
        $translation->setTranslatable($this);

        return $this;
    }

    /**
     * Removes specific translation.
     *
     * @param Translation $translation The translation
     */
    public function removeTranslation($translation)
    {
        $this->getTranslations()->removeElement($translation);
    }

    /**
     * Returns translation for specific locale (creates new one if doesn't exists).
     * If requested translation doesn't exist, it will first try to fallback default locale
     * If any translation doesn't exist, it will be added to newTranslations collection.
     * In order to persist new translations, call mergeNewTranslations method, before flush
     *
     * @param string $locale The locale (en, ru, fr) | null If null, will try with current locale
     *
     * @return Translation
     */
    public function translate($locale = null, $fallbackToDefault = true, $fallbackLocale = 'en')
    {
        return $this->doTranslate($locale, $fallbackToDefault, $fallbackLocale);
    }

    protected function doTranslate($locale = null, $fallbackToDefault = true, $fallbackLocale = 'en')
    {
        if (null === $locale) {
            $locale = $this->getCurrentLocale();
        }

        $translation = $this->findTranslationByLocale($locale);
        if ($translation and !$translation->isEmpty()) {
            return $translation;
        }

        if ($fallbackToDefault && $defaultTranslation = $this->findTranslationByLocale($fallbackLocale, false)) {
            return $defaultTranslation;
        }

        $class       = self::getTranslationEntityClass();
        $translation = new $class();
        $translation->setLocale($locale);

        $this->getNewTranslations()->set($translation->getLocale(), $translation);
        $translation->setTranslatable($this);

        return $translation;
    }

    /**
     * Merges newly created translations into persisted translations.
     */
    public function mergeNewTranslations()
    {
        foreach ($this->getNewTranslations() as $newTranslation) {
            if (!$this->getTranslations()->contains($newTranslation)) {
                $this->addTranslation($newTranslation);
                $this->getNewTranslations()->removeElement($newTranslation);
            }
        }
    }

    /**
     * @param mixed $locale the current locale
     */
    public function setCurrentLocale($locale)
    {
        $this->currentLocale = $locale;
    }

    public function getCurrentLocale()
    {
        return $this->currentLocale ?: $this->getDefaultLocale();
    }

    public function getDefaultLocale()
    {
        return 'en';
    }

    protected function proxyCurrentLocaleTranslation($method, array $arguments = [], $fallbackLocale = 'en')
    {
        return call_user_func_array(
            [$this->translate($this->getCurrentLocale(), true, $fallbackLocale), $method],
            $arguments
        );
    }

    /**
     * Returns translation entity class name.
     *
     * @return string
     */
    public static function getTranslationEntityClass()
    {
        return __CLASS__.'Translation';
    }

    /**
     * Finds specific translation in collection by its locale.
     *
     * @param string $locale              The locale (en, ru, fr)
     * @param bool   $withNewTranslations searched in new translations too
     *
     * @return Translation|null
     */
    protected function findTranslationByLocale($locale, $withNewTranslations = true)
    {
        $translation = $this->getTranslations()->get($locale);

        if ($translation) {
            return $translation;
        }

        if ($withNewTranslations) {
            return $this->getNewTranslations()->get($locale);
        }
    }
}

```

**Comment by MAXakaWIZARD** [comment]
Me also encountered that problem, I need different default locale.

**Comment by docteurklein** [comment]
Juste override the method then?

**Comment by charlieWasp** [comment]
@docteurklein How?

**Comment by docteurklein** [comment]
in the class that uses the translatable trait, define your own `getDefaultLocale` method.

**Comment by charlieWasp** [comment]
@docteurklein of course! At the end of the day I might ask dumb questions :) As a quick fix it's a good solution

**Comment by markitosgv** [comment]
Yeah its for now easiest solution to this, but i think we need to think some way to be configurable in DoctrineBehaviours

**Comment by kuczek** [comment]
Yes, I agree, there should be a way to define default locales for all classess in one place.

**Comment by markitosgv** [comment]
If you are using symfony is possible to override a trait like controller or asset?

**Comment by kuczek** [comment]
@markitosgv It's not possible, is not based on DI system, but on PHP mechanism, which do not allow you to override existing trait. Only solution is to redefine autoloading system, but it's to much to do for such thing.

**Comment by Einenlum** [comment]
Solved in #191 ?
